Map and implicit type

The following program `sample.go` triggers an error:

```go
package main

import (
	"fmt"
	"sort"
)

type Foo struct {
	Name string
}

func main() {
	m := map[string]Foo{
		"hello": { Name: "bar"},
		"world": { Name: "machin"},
	}

	var content []string

	for key, value := range m {
		content = append(content, key+value.Name)
	}

	sort.Strings(content)
	fmt.Println(content)
}
```

Expected result:
```console
$ go run ./sample.go
[hellobar worldmachin]
```

Got:
```console
$ dyngo ./sample.go
./_test/map8.go:14:13: type definition not implemented: KeyValueExpr
```

--- a/_test/map12.go
+++ b/_test/map12.go
@@ -1,0 +1,29 @@
+package main
+
+import (
+	"fmt"
+	"sort"
+)
+
+type Foo struct {
+	Name string
+}
+
+func main() {
+	m := map[string]Foo{
+		"hello": {Name: "bar"},
+		"world": {Name: "machin"},
+	}
+
+	var content []string
+
+	for key, value := range m {
+		content = append(content, key+value.Name)
+	}
+
+	sort.Strings(content)
+	fmt.Println(content)
+}
+
+// Output:
+// [hellobar worldmachin]

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -516,13 +516,14 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 				n.gen = mapLit
 			case StructT:
 				n.action, n.gen = CompositeLit, compositeLit
+				child := n.child
+				if child[0].isType(scope) {
+					child = n.child[1:]
+				}
 				// Handle object assign from sparse key / values
-				if len(n.child) > 1 && n.child[1].kind == KeyValueExpr {
+				if len(child) > 0 && child[0].kind == KeyValueExpr {
 					n.gen = compositeSparse
-					if n.typ, err = nodeType(interp, scope, n.child[0]); err != nil {
-						return
-					}
-					for _, c := range n.child[1:] {
+					for _, c := range child {
 						c.findex = n.typ.fieldIndex(c.child[0].ident)
 					}
 				}
@@ -1031,7 +1032,7 @@ func (n *Node) isType(scope *Scope) bool {
 		if p, ok := n.interp.binType[pkg]; ok && p[name] != nil {
 			return true // Imported binary type
 		}
-		if p, ok := n.interp.scope[pkg]; ok && p.sym[name] != nil {
+		if p, ok := n.interp.scope[pkg]; ok && p.sym[name] != nil && p.sym[name].kind == Typ {
 			return true // Imported source type
 		}
 	case Ident:

--- a/interp/interp_test.go
+++ b/interp/interp_test.go
@@ -2790,6 +2790,46 @@ func main() {
 	// [hellobar hellofoo worldmachin worldtruc]
 }
 
+func Example_map12() {
+	src := `
+package main
+
+import (
+	"fmt"
+	"sort"
+)
+
+type Foo struct {
+	Name string
+}
+
+func main() {
+	m := map[string]Foo{
+		"hello": {Name: "bar"},
+		"world": {Name: "machin"},
+	}
+
+	var content []string
+
+	for key, value := range m {
+		content = append(content, key+value.Name)
+	}
+
+	sort.Strings(content)
+	fmt.Println(content)
+}
+`
+	i := interp.New(interp.Opt{Entry: "main"})
+	i.Use(stdlib.Value, stdlib.Type)
+	_, err := i.Eval(src)
+	if err != nil {
+		panic(err)
+	}
+
+	// Output:
+	// [hellobar worldmachin]
+}
+
 func Example_map2() {
 	src := `
 package main

--- a/interp/run.go
+++ b/interp/run.go
@@ -848,13 +848,11 @@ func _return(n *Node) {
 func arrayLit(n *Node) {
 	value := valueGenerator(n, n.findex)
 	next := getExec(n.tnext)
-	var child []*Node
-
-	if n.typ.untyped {
-		child = n.child
-	} else {
+	child := n.child
+	if !n.typ.untyped {
 		child = n.child[1:]
 	}
+
 	a, _ := n.typ.zero()
 	values := make([]func(*Frame) reflect.Value, len(child))
 	for i, c := range child {
@@ -886,7 +884,10 @@ func arrayLit(n *Node) {
 func mapLit(n *Node) {
 	value := valueGenerator(n, n.findex)
 	next := getExec(n.tnext)
-	child := n.child[1:]
+	child := n.child
+	if !n.typ.untyped {
+		child = n.child[1:]
+	}
 	typ := n.typ.TypeOf()
 	keys := make([]func(*Frame) reflect.Value, len(child))
 	values := make([]func(*Frame) reflect.Value, len(child))
@@ -911,13 +912,11 @@ func mapLit(n *Node) {
 func compositeLit(n *Node) {
 	value := valueGenerator(n, n.findex)
 	next := getExec(n.tnext)
-	var child []*Node
-
-	if n.typ.untyped {
-		child = n.child
-	} else {
+	child := n.child
+	if !n.typ.untyped {
 		child = n.child[1:]
 	}
+
 	a, _ := n.typ.zero()
 	values := make([]func(*Frame) reflect.Value, len(child))
 	for i, c := range child {
@@ -938,7 +937,11 @@ func compositeLit(n *Node) {
 func compositeSparse(n *Node) {
 	value := valueGenerator(n, n.findex)
 	next := getExec(n.tnext)
-	child := n.child[1:]
+	child := n.child
+	if !n.typ.untyped {
+		child = n.child[1:]
+	}
+
 	values := make(map[int]func(*Frame) reflect.Value)
 	a, _ := n.typ.zero()
 	for _, c := range child {

